### PR TITLE
Make wxDC and wxGraphicsContext DPI aware

### PR DESCRIPTION
**Pull request by MaartenBent**
Add and use `FromDIP` and `ToDIP`, and make returned values from `GetDPI`/`GetPPI` consistent with `wxWindow`.

I am testing this mainly with the drawing sample (polygon page) and printing sample which have been made DPI aware. I am checking if all renderers, saved files, print to pdf, look the same compared to default 96dpi/100%.

This includes:
* msw
  * wxDC, wxGCDC with gdi+, d2d, and cairo
  * with and without wxBufferedPaintDC
  * saving as bmp, svg, ps
  * at different dpi
    * static: main monitor with dpi other dan 96/100%
    * dynamic: move from one dpi to another
  * different dpi in manifest: no, system, per-monitor

* osx
  * wxDC, wxGCDC with cg and cairo
  * with and without wxBufferedPaintDC
  * saving as bmp
  * at 144dpi/retina and with 'Open in Low Resolution' option
 
* gtk3
  * wxDC, wxGCDC with cairo
  * with and without wxBufferedPaintDC
  * saving as bmp
  * with scale 100% and 200%

---
TODO
- [x] Check gtk3
- [x] Update interface files (remove `GetDPIScaleFactor`, add `FromDIP`, `ToDIP`)

### CODE REVIEW

File: src/msw/dcmemory.cpp [hunk 1]
```diff
@@ -143,9 +143,12 @@ void wxMemoryDCImpl::DoSelect( const wxBitmap& bitmap )
...
 }
 
-double wxMemoryDCImpl::GetDPIScaleFactor() const
+wxSize wxMemoryDCImpl::GetPPI() const
 {
-    return m_contentScaleFactor;
+    if ( GetWindow() )
+        return GetWindow()->GetDPI();
+    else
+        return wxDisplay::GetStdPPI() * GetContentScaleFactor();
```
Comment (vadz) [inline src/msw/dcmemory.cpp:151]: Sorry, I don't understand this:

1. How can a `wxMemoryDC` have an associated window?
2. Why do we need multiply only the display DPI but not the window one by the scale factor?

File: src/msw/dcprint.cpp [hunk 2]
```diff
@@ -230,6 +231,13 @@ wxRect wxPrinterDCImpl::GetPaperRect() const
     return wxRect(x, y, w, h);
 }
 
+void wxPrinterDCImpl::SetFont(const wxFont& font)
+{
+    wxFont scaledFont = font;
+    if ( scaledFont.IsOk() )
+        scaledFont.WXAdjustToPPI(wxDisplay::GetStdPPI());
```
Comment (vadz) [inline src/msw/dcprint.cpp:238]: I'm not sure why is this needed neither...

File: src/msw/enhmeta.cpp [hunk 3]
```diff
@@ -223,6 +224,24 @@ class wxEnhMetaFileDCImpl : public wxMSWDCImpl
...
+    virtual wxSize ToDIP(const wxSize& sz) const wxOVERRIDE
+    {
+        return sz;
+    }
+
+    void SetFont(const wxFont& font) wxOVERRIDE
+    {
+        wxFont scaledFont = font;
+        if (scaledFont.IsOk())
+            scaledFont.WXAdjustToPPI(wxDisplay::GetStdPPI());
```
Comment (vadz) [inline src/msw/enhmeta.cpp:241]: Same question, why do we need to do this?

File: src/generic/graphicc.cpp [hunk 4]
```diff
@@ -1124,15 +1130,22 @@ wxCairoFontData::InitFontComponents(const wxString& facename,
...
-                         const wxColour& col )
+                                  const wxRealPoint& dpi, const wxColour& col )
     : wxGraphicsObjectRefData(renderer)
 #ifdef __WXGTK__
     , m_wxfont(font)
 #endif
 {
     InitColour(col);
 
+#ifndef __WXMAC__
```
Comment (vadz) [inline src/generic/graphicc.cpp:1141]: I'd prefer to test for `__WXMAC__` rather than its negation, it's more clear and more consistent with the existing tests below.

File: src/generic/graphicc.cpp [hunk 5]
```diff
@@ -1124,15 +1130,22 @@ wxCairoFontData::InitFontComponents(const wxString& facename,
...
+    m_size = !dpi.y
+        ? double(font.GetPixelSize().GetHeight())
+        : double(font.GetFractionalPointSize() * dpi.y / 72);
+#else
+    wxUnusedVar(dpi);
     m_size = font.GetPointSize();
```
Comment (vadz) [inline src/generic/graphicc.cpp:1147]: Why not `GetFractionalPointSize()`?

File: src/generic/graphicc.cpp [hunk 6]
```diff
@@ -1124,15 +1130,22 @@ wxCairoFontData::InitFontComponents(const wxString& facename,
...
 #ifdef __WXGTK__
     , m_wxfont(font)
 #endif
 {
     InitColour(col);
 
+#ifndef __WXMAC__
+    m_size = !dpi.y
+        ? double(font.GetPixelSize().GetHeight())
+        : double(font.GetFractionalPointSize() * dpi.y / 72);
```
Comment (vadz) [inline src/generic/graphicc.cpp:1144]: Could you please explain why do we do this and why do we do a different thing under Mac?

This must be hiding some more fundamental incompatibility/difference between the platforms, but I don't know which one.

File: src/generic/graphicc.cpp [hunk 7]
```diff
@@ -2749,6 +2762,31 @@ void wxCairoContext::Flush()
...
+            *dpiX = dpi.x;
+        if ( dpiY )
+            *dpiY = dpi.y;
+    }
+    else
+    {
+        double dpi = (double)wxDisplay::GetStdPPIValue();
+#ifndef __WXMAC__
+        dpi *= GetContentScaleFactor();
+#endif
```
Comment (vadz) [inline src/generic/graphicc.cpp:2781]: Again, this seems to indicate that there is something different with Cairo under Mac, but what?

**Review by vadz — COMMENTED** [review]
Thanks a lot for all the work here!

It would be great if you could please answer my various questions below -- sorry if I'm missing something obvious, but some of the changes look puzzling (which doesn't mean that they're incorrect, of course). TIA!

File: samples/printing/printing.cpp [hunk 8]
```diff
@@ -142,79 +142,79 @@ void MyApp::Draw(wxDC&dc)
     dc.SetPen(*wxBLACK_PEN);
     dc.SetBrush(*wxLIGHT_GREY_BRUSH);
 
-    dc.DrawRectangle(0, 0, 230, 350);
-    dc.DrawLine(0, 0, 229, 349);
-    dc.DrawLine(229, 0, 0, 349);
+    dc.DrawRectangle(dc.FromDIP(0), dc.FromDIP(0), dc.FromDIP(230), dc.FromDIP(350));
```
[the remark this answers is not in the document]
Comment (csomor) [reply samples/printing/printing.cpp:145]: would having some global version like `wxFromDIP` and `wxToDIP` with overloads taking wxPoints and wxRectangles (and returning the same type but converted) and a second parameter for wxDC, wxGraphicsContext etc. help to make things more readable ?

File: src/generic/graphicc.cpp [hunk 9]
```diff
@@ -2237,7 +2250,7 @@ wxCairoContext::wxCairoContext( wxGraphicsRenderer* renderer, const wxMemoryDC&
 #endif
 
 #ifdef __WXMAC__
-    CGContextRef cgcontext = (CGContextRef)dc.GetWindow()->MacGetCGContextRef();
+    CGContextRef cgcontext = (CGContextRef)dc.GetSelectedBitmap().GetHBITMAP();
```
Comment (csomor) [inline src/generic/graphicc.cpp:2253]: what is the reason for this change ? IIRC this bitmap-context does not know about its content scaling

**Comment by csomor** [comment]
Thanks a lot @MaartenBent for accepting this challenge!

File: src/generic/graphicc.cpp (hunk 6 above)
Comment (MaartenBent) [reply src/generic/graphicc.cpp:1144]: I think there are two differences here between macOS and Windows.

The first one is that macOS is based on 72dpi and Windows on 96dpi. Fonts are in 72ppi, so on Windows they have to be corrected to the DPI. On macOS no correction is needed.

Second, on Windows the font has to be adjusted to the screen DPI (96 at 100%, 192 at 200%) to get the correct font size. On macOS this is not needed (content scale stuff?) so you basically end up with `font.GetFractionalPointSize() * 72 / 72)`.

File: src/generic/graphicc.cpp (hunk 5 above)
Comment (MaartenBent) [reply src/generic/graphicc.cpp:1147]: Yes, that can be used here.

File: src/generic/graphicc.cpp (hunk 9 above)
Comment (MaartenBent) [reply src/generic/graphicc.cpp:2253]: Thanks, this works too (but saved images are the same, so maybe the bitmap-context does know about it).

File: build/cmake/init.cmake [hunk 10]
```diff
@@ -574,6 +574,17 @@ if(wxUSE_GUI)
...
         endif()
     endif()
+
+    if(wxUSE_CAIRO AND NOT WXGTK)
+        find_package(Cairo)
+        if(NOT CAIRO_FOUND)
+            message(WARNING "Cairo not found, Cairo renderer won't be available")
+            wx_option_force_value(wxUSE_CAIRO OFF)
+        else()
+            #TODO only if build-in png?
```
[the remark this answers is not in the document]
Comment (MaartenBent) [reply build/cmake/init.cmake:584]: This `else` is obsolete now and can be removed. It was for a different `FindCairo.cmake` that included all cairo dependent headers (lzo, libpng16, freetype2, harfbuzz, glib-2.0, pixman-1), which could cause problems if `imagpng.cpp` used the wrong png header.
The current one only adds `path/include/cairo` which is enough to build.

File: src/generic/graphicc.cpp (hunk 6 above)
Comment (vadz) [reply src/generic/graphicc.cpp:1144]: I'd prefer to use `wxDisplay::GetStdPPIValue() / 72`, this should be optimized away under Mac and would result in more clear code IMO.

As for "DPI correction", I'd indeed expect it to be taken care of by `GetContentScaleFactor()`, but its value is the same for GTK and macOS, while this code is Mac-specific. I.e. it's not so much the difference between MSW and Mac that I don't understand, but the difference between Mac and GTK.

Comment (MaartenBent) [reply src/generic/graphicc.cpp:1144]: I tried both on GTK and the font looks exactly the same. Looking further into this, it is because GTK uses pango for font drawing. `m_size` isn't even used, it uses `m_wxfont`.

File: samples/drawing/drawing.cpp [hunk 11]
```diff
@@ -1999,8 +2000,8 @@ void MyCanvas::Draw(wxDC& pdc)
     {
         wxCoord x0, y0;
         dc.GetDeviceOrigin(&x0, &y0);
-        m_sizeX = dc.LogicalToDeviceX(dc.MaxX()) - x0 + 1;
-        m_sizeY = dc.LogicalToDeviceY(dc.MaxY()) - y0 + 1;
+        m_sizeX = dc.ToDIP(dc.LogicalToDeviceX(dc.MaxX()) - x0) + 1;
```
[the remark this answers is not in the document]
Comment (MaartenBent) [reply samples/drawing/drawing.cpp]: It's not only `wxSVGFileDC`, but `CreateWithDIPSize` also needs a size in DIP.
I updated the member and function names.

File: samples/drawing/drawing.cpp [hunk 12]
```diff
@@ -2012,7 +2013,7 @@ void MyCanvas::OnMouseMove(wxMouseEvent &event)
         PrepareDC(dc);
         m_owner->PrepareDC(dc);
 
-        wxPoint pos = dc.DeviceToLogical(event.GetPosition());
+        wxPoint pos = dc.ToDIP(dc.DeviceToLogical(event.GetPosition()));
```
[the remark this answers is not in the document]
Comment (MaartenBent) [reply samples/drawing/drawing.cpp]: I updated it to show both logical coordinates and wxDC DIP coordinates (only when different).

File: samples/printing/printing.cpp [hunk 13]
```diff
@@ -111,8 +111,8 @@ bool MyApp::OnInit(void)
     // ----------------------------
 
     MyFrame* frame = new MyFrame((wxFrame *) NULL, "wxWidgets Printing Demo",
-                                 wxPoint(0, 0), wxSize(400, 400));
-
+                                 wxDefaultPosition, wxDefaultSize);
+    frame->SetSize(frame->FromDIP(wxSize(400, 400)));
```
[the remark this answers is not in the document]
Comment (MaartenBent) [reply samples/printing/printing.cpp]: I updated the constructors in the drawing sample and printing sample so these unused arguments are not needed anymore.

File: samples/printing/printing.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (MaartenBent) [reply samples/printing/printing.cpp]: I used `FromDIP(0)` for consistency, removed it now.

I don't think adding more dip functions while make it better readable. Maybe a `DrawRectangle` that just accepts a rectangle to reduce the number of `FromDIP` (like `dc.DrawRectangle(dc.FromDIP(wxRect(0, 0, 230, 350)))`) but I wont do that in this PR.

File: samples/printing/printing.cpp [hunk 14]
```diff
@@ -392,7 +392,7 @@ void MyFrame::OnPrintPreview(wxCommandEvent& WXUNUSED(event))
     }
 
     wxPreviewFrame *frame =
-        new wxPreviewFrame(preview, this, "Demo Print Preview", wxPoint(100, 100), wxSize(600, 650));
+        new wxPreviewFrame(preview, this, "Demo Print Preview", wxPoint(100, 100), FromDIP(wxSize(600, 650)));
```
[the remark this answers is not in the document]
Comment (MaartenBent) [reply samples/printing/printing.cpp]: For a position it was not that important, so I left `FromDIP` out. Changed it to `wxDefaultPosition` now.

File: src/msw/dcmemory.cpp (hunk 1 above)
Comment (MaartenBent) [reply src/msw/dcmemory.cpp]: I added this for a workaround I mentioned before (using `memdc.GetImpl()->SetWindow( )`). And that window already has the correct DPI so no scaling would be needed.

It is probably better to remove it, so I did.

File: src/generic/graphicc.cpp (hunk 6 above)
Comment (MaartenBent) [reply src/generic/graphicc.cpp]: I've restructured `wxCairoFontData` constructor and added some comments, to address this and above comments.

File: src/msw/dcprint.cpp (hunk 2 above)
Comment (MaartenBent) [reply src/msw/dcprint.cpp:238]: Based on testing I found that `wxPrinterDCImpl` and `wxEnhMetaFileDCImpl` below are based on 96DPI. So the font has to be adjusted to this DPI as well.
Otherwise the font size in the print preview and print (to pdf) are too big.

File: src/generic/graphicc.cpp [hunk 15]
```diff
@@ -1124,20 +1130,30 @@ wxCairoFontData::InitFontComponents(const wxString& facename,
...
-#elif defined(__WXGTK__)
+#ifdef __WXMSW__
+    // Font is 72 DPI, screen is 96-based, and font needs a correction
+    // for screens with higher DPI (similar to gdi+ and d2d renderers).
+    m_size = !dpi.y
+    ? double(font.GetPixelSize().GetHeight())
+    : (font.GetFractionalPointSize() * dpi.y / 72);
 #else
+    // On macOS, font and screen both are 72 DPI, an macOS font size does
+    // not need a correction for retina displays. GTK does not use m_size.
```
Comment (vadz) [inline src/generic/graphicc.cpp:1149]: Should we perhaps explain in this comment why do we still set `m_size` for it then? Or perhaps not set it, i.e. replace `#else` above with `#elif defined(__WXMAC__)`?

File: samples/drawing/drawing.cpp [hunk 16]
```diff
@@ -1896,15 +1887,18 @@ void MyCanvas::Draw(wxDC& pdc)
     }
 
     if ( m_clip )
-        dc.SetClippingRegion(100, 100, 100, 100);
+    {
+        int clipSize = dc.FromDIP(100);
```
Comment (vadz) [inline samples/drawing/drawing.cpp:1891]: Sorry for nitpicking, but this is not a sizer, is it? I.e. it's used also as `x` and `y` arguments.

**Review by vadz — APPROVED** [review]
Looks perfect to me, thanks a lot!

I can just tweak the comment and the variable name mentioned below myself and merge already if you'd like, please let me know.

And thanks again!

**Comment by MaartenBent** [comment]
I still need to do the interface files, so I can address the comments as well.

**Comment by vadz** [comment]
Thanks a lot once again, AFAICS this should be merged and I'll do it soon unless @csomor has any comments/objections.

What would be a good way to briefly summarize CMake-related changes in this PR in the merge commit message?

**Comment by MaartenBent** [comment]
> What would be a good way to briefly summarize CMake-related changes in this PR in the merge commit message?

Add support for finding Cairo when not using GTK, fix a warning when creating the inplace-config, and mark some variables as advanced.
